src/Interpreters/DDLWorker.cpp

```diff
@@ -338,10 +338,10 @@ void DDLWorker::scheduleTasks()
     if (!server_startup)
     {
         /// We will recheck status of last executed tasks. It's useful if main thread was just restarted.
-        auto & min_task = *std::min_element(current_tasks.begin(), current_tasks.end());
+        auto & min_task = current_tasks.front();
         String min_entry_name = last_skipped_entry_name ? std::min(min_task->entry_name, *last_skipped_entry_name) : min_task->entry_name;
         begin_node = std::upper_bound(queue_nodes.begin(), queue_nodes.end(), min_entry_name);
-        current_tasks.clear();
+        current_tasks.remove_if([](const DDLTaskPtr & t) { return t->completely_processed.load(); });
```

Comment:
Why should we save tasks which was not executed? If task was executed, but its status was not written to ZooKeeper (due to connection loss), then we call `processTask(...)` [here](https://github.com/ClickHouse/ClickHouse/blob/6946514c0b3165ab79de7a9feca5d19d0a73c03e/src/Interpreters/DDLWorker.cpp#L321) to write the status. If task was not executed, then we will read the task from ZooKeeper again and retry it.

Reply:
Reference to current_tasks is used in processTasks, in case of distirbuted_ddl.pool_size > 1 they are processed in background and such items cannot be removed from the current_tasks.

But this does not solve iterating over those tasks again, seems that some pending state should be introduced for this async execution, what do you think?

Reply:
> Reference to current_tasks is used in processTasks

You're right, I have removed `reset_state()` from [here](https://github.com/ClickHouse/ClickHouse/blob/6946514c0b3165ab79de7a9feca5d19d0a73c03e/src/Interpreters/DDLWorker.cpp#L1009), but forgot to add something like `if (worker_pool) worker_pool->wait()` for the case when `pool_size > 1`.

> But this does not solve iterating over those tasks again, seems that some pending state should be introduced for this async execution, what do you think?

I think that we should not have any background tasks running if main thread was restarted/reinitialied due to connection loss or some unexpected exception.

Reply:
>You're right, I have removed reset_state() from here, but forgot to add something like if (worker_pool) worker_pool->wait() for the case when pool_size > 1.

Yes, this also should be addressed.

>I think that we should not have any background tasks running if main thread was restarted/reinitialied due to connection loss or some unexpected exception.

If the main thread was restarted then yes, but that code called every time not only after restarting the main thread, am I missing something? (and also the code above that iterates over current_tasks too)